src/day22.scala

```diff
@@ -0,0 +1,114 @@
+
+package day22
+
+import scala.util.Using
+import scala.io.Source
+
+import Command.*
+import scala.collection.mutable.ListBuffer
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day22"))(_.mkString)
+
+case class Dimension(min: Int, max: Int):
+  require(min <= max)
+
+  def isSubset(d: Dimension): Boolean =
+    min >= d.min && max <= d.max
+
+  infix def insersect(d: Dimension): Option[Dimension] =
+    Option.when(max >= d.min && min <= d.max) {
+      (min max d.min) by (max min d.max)
+    }
+
+  def size: Int = max - min + 1
+
+extension (x1: Int)
+  infix def by (x2: Int): Dimension = Dimension(x1, x2)
+
+case class Cuboid(xs: Dimension, ys: Dimension, zs: Dimension):
+  def volume: BigInt = BigInt(xs.size) * ys.size * zs.size
+
+enum Command:
+  case On, Off
+
+case class Step(command: Command, cuboid: Cuboid)
+
+case class Cube(x: Int, y: Int, z: Int)
+
+def intersect(old: Cuboid, curr: Cuboid): Option[Cuboid] =
+  for
+    xs <- old.xs insersect curr.xs
+    ys <- old.ys insersect curr.ys
+    zs <- old.zs insersect curr.zs
+  yield
+    Cuboid(xs, ys, zs)
+
+def subdivide(old: Cuboid, hole: Cuboid): Set[Cuboid] =
+  var divisions = Set.empty[Cuboid]
+  if old.xs.min != hole.xs.min then
+    divisions += Cuboid(xs = old.xs.min by hole.xs.min - 1, ys = old.ys, zs = old.zs)
+  if old.xs.max != hole.xs.max then
+    divisions += Cuboid(xs = hole.xs.max + 1 by old.xs.max, ys = old.ys, zs = old.zs)
+  if old.ys.min != hole.ys.min then
+    divisions += Cuboid(xs = hole.xs, ys = old.ys.min by hole.ys.min - 1, zs = old.zs)
+  if old.ys.max != hole.ys.max then
+    divisions += Cuboid(xs = hole.xs, ys = hole.ys.max + 1 by old.ys.max, zs = old.zs)
+  if old.zs.min != hole.zs.min then
+    divisions += Cuboid(xs = hole.xs, ys = hole.ys, zs = old.zs.min by hole.zs.min - 1)
+  if old.zs.max != hole.zs.max then
+    divisions += Cuboid(xs = hole.xs, ys = hole.ys, zs = hole.zs.max + 1 by old.zs.max)
+  divisions
+
+def run(steps: Iterator[Step]): Set[Cuboid] =
+  steps.foldLeft(Set.empty)((on, step) =>
+    on.foldLeft(if step.command == On then Set(step.cuboid) else Set.empty)((newOn, old) =>
+      intersect(old, step.cuboid) match
+        case Some(hole) =>
+          newOn | subdivide(old, hole)
+        case _ =>
+          newOn + old
```

Comment:
I find it hard to read a `foldLeft` inside a `foldLeft`. So I suggest to extract this as method with a proper name, it could be a method of `Step` maybe.